### ISSUE TITLE
Create a global delay_factor

### ISSUE THREAD

**Issue by ktbyers** [ktbyers/netmiko#98]

**Comment by jfilliben** [comment]
Is there a way to pass a timeout/delay value to ConnectHandler? I am attempting to manage some slow IOS devices on the other side of the globe and I believe the default 0.5 second delay is causing me issues:

Traceback (most recent call last):
  File "dns_crawl.py", line 159, in <module>
    interfacesDict[router] = get_interfaces(routerDict)
  File "dns_crawl.py", line 66, in get_interfaces
    with ssh_manager(routerDict) as netConnect:
  File "C:\Users\308001282\AppData\Local\Continuum\Anaconda2\lib\contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "dns_crawl.py", line 23, in ssh_manager
    conn = SSHClient(**net_device)
  File "build\bdist.win-amd64\egg\netmiko\base_connection.py", line 45, in __init__
  File "build\bdist.win-amd64\egg\netmiko\base_connection.py", line 61, in session_preparation
  File "build\bdist.win-amd64\egg\netmiko\base_connection.py", line 201, in set_base_prompt
ValueError: Router prompt not found: e

**Comment by ktbyers** [comment]
@jfilliben I added a new global_delay_factor argument, you use it when you call the ConnectHandler as follows:

cisco_881 = {
   'device_type': 'cisco_ios',
   'ip': ‘1.1.1.1’,
   'username': ‘admin’,
   'password': ‘password’,
   'global_delay_factor': 3,
}

net_connect = ConnectHandler(**cisco_881)

Netmiko will choose the larger of the two (global_delay_factor or the delay_factor specified in an individual method). By default, they usually are set the same at a value of .5 second.

You will need to use the master branch on GitHub to use the code.

Let me know if this doesn't fix your issue (or isn't working properly).

**Comment by jfilliben** [comment]
Awesome, thanks! I'll test it out and let you know if it helps on Monday.

**Comment by jfilliben** [comment]
Kirk,

I couldn't wait to test :)  I tried it out this morning and it worked perfectly... more slowly, of course, but it worked through all the distant routers without error.

Now I need to work on parallel processing to cut down on the run time.

Thanks!
Jeremy

**Comment by ktbyers** [comment]
Nice :-)

Either using threads or parallel processes should be fine. Python's GIL
(global interpreter lock) won't be an issue here since so much of time is
spent waiting on IO.

Here is an example using Netmiko + Threads:
https://github.com/ktbyers/pynet/blob/master/pyth_ans_ecourse/class8/ex6_threads_show_ver.py

Here is an example using Netmiko + Multiple processes:
https://github.com/ktbyers/pynet/blob/master/pyth_ans_ecourse/class8/ex7_processes_show_ver.py


Let me know if you need any assistance on the parallel processing.


Regards,

Kirk

On Sat, Nov 14, 2015 at 8:53 AM, Jeremy Filliben <notifications@github.com>
wrote:

> Kirk,
>
> I couldn't wait to test :) I tried it out this morning and it worked
> perfectly... more slowly, of course, but it worked through all the distant
> routers without error.
>
> Now I need to work on parallel processing to cut down on the run time.
>
> Thanks!
> Jeremy
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ktbyers/netmiko/issues/98#issuecomment-156719444>.
>



-- 
Kirk Byers
CCIE #6243 emeritus
ktbyers@twb-tech.com
*http://pynet.twb-tech.com <http://pynet.twb-tech.com/blog>*
(415) 264-6270
Simplify your network management through automation

**Comment by sivabudh** [comment]
@ktbyers I'm new to `netmiko` and after reading this, is my understanding correct?

* It's possible to set a default timeout by setting a  `global_delay_factor` before passing the hash to `ConnectHandler`.
* It's also possible to parallelize SSH connections to multiple network devices by following threads / multiple processes examples above.

**Comment by ktbyers** [comment]
@sivabudh global_delay_factor is not a timeout; it controls the delay for various operations. Most of the time you shouldn't need to set this, but some devices are really slow (particularly virtual devices) so you need to tell Netmiko to add more delay into its operations.

You can also generally override specific commands and increase the delay for just that command (if you have a long running command).

You can run the SSH connections concurrently using threads or multiprocesses (as per the above examples). So you can fire off 10 SSH sessions and the operation will complete in a similar time as one SSH connection would have completed.

**Comment by annejg** [comment]
Hi @ktbyers 

I noticed your comment above:  

> You can also generally override specific commands and increase the delay for just that command (if you have a long running command).

Is this doable only by wrapping a particular command in the ConnectHandler session or can I add delay to specific commands? for example when using ConnectHandler: 

`net_connect.send_command('clear` configuration commits diskspace 2000\n') net_connect.send_command('\n')`

(these commands are inside a function with other commands that do not need delay). 

Thank You in advance!

**Comment by ktbyers** [comment]
@annejg You can do something like the following,

```
net_connect.send_command('clear configuration commits diskspace 2000', delay_factor=2)
```

That will double the delays of that one particular send_command execution.

Note, if the command prompts you for something like "Do you want to continue?". Then you typically need to do something like:

```
output = net_connect.send_command_timing("some command that prompts")
if 'Do you want to continue' in output:
    output += net_connect.send_command_timing("y")
```

**Comment by annejg** [comment]
@ktbyers Thank you very much!  

I added it as such: 

`net_connect.send_command('clear configuration commits diskspace 2000\n)`
`net_connect.send_command('\n', delay_factor=2)`

and it works perfect!

**Comment by ciscolabs101** [comment]
After so many years this is so good and works like a charm. Thanks @ktbyers

**Comment by ajuarnair** [comment]
My cod is to just write the configuration and take a backup into TFTP server.But it says "Time out error-Increase the time out to lareger value.I tried to add read timeout and globla factor but in vain".Below is the code. Can you please add if am missing something.

[bkup.txt](https://github.com/ktbyers/netmiko/files/11272962/bkup.txt)

from netmiko import ConnectHandler  
device = {     'device_type': 'cisco_ios',     "ip": "10.154.12.11",     "username": "johnconnor",     "password": "@merica1776",  }  net_connect = ConnectHandler(**device)  
net_connect.enable()  
output = net_connect.send_config_set('do write', read_timeout=90.0)   
Tftp_server = "10.0.34.42" 
cmd1 = "copy running-config tftp:" 
result = net_connect.send_command_timing(cmd1,) 
if "Address or name of remote host" in result:  
   result += net_connect.send_command_timing(Tftp_server)  
if 'Destination filename' in result:   
  result += net_connect.send_command_timing("backup11.txt")  
print(output) 
print(cmd1) 
net_connect.disconnect()

**Comment by ktbyers** [comment]
@ajuarnair This issue is several years old and closed, you should create a new issue if you are having a problem.

Please ensure that you format your markdown correctly when creating the issue.
